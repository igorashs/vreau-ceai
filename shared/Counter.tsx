import styled from 'styled-components';
import Button from '@/shared/Button';
import MinusSvg from '@/icons/minus.svg';
import PlusSvg from '@/icons/plus.svg';

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  gap: 7px;
`;

const CountInput = styled.input`
  border: 0;
  width: calc(var(--baseline) * 2);
  background-color: transparent;
  color: var(--accent-text-dark);
  border-bottom: 1px solid var(--accent);
  text-align: center;

  /* Chrome, Safari, Edge, Opera */
  ::-webkit-outer-spin-button,
  ::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }

  /* Firefox */
  &[type='number'] {
    -moz-appearance: textfield;
  }
`;

type CounterProps = {
  count?: number;
  min?: number;
  max?: number;
  onChange: (value: string | number) => void;
  onDecrement?: (value: string | number) => void;
  onIncrement?: (value: string | number) => void;
};

const Counter = ({
  count = 0,
  min = 0,
  max = Infinity,
  onChange,
  onDecrement = onChange,
  onIncrement = onChange,
}: CounterProps) => {
  return (
    <Wrapper>
      <Button
        icon
        noPadding
        onClick={() => onDecrement(count > min ? count - 1 : count)}
      >
        <MinusSvg />
      </Button>
      <CountInput
        type="number"
        value={count}
        onBlur={(e) => !e.currentTarget.value && onChange(min)}
        onFocus={(e) => e.currentTarget.select()}
        onChange={(e) => {
          if (!e.currentTarget.value) {
            onChange('');
          } else {
            const value = +e.currentTarget.value;

            if (value >= min && value <= max) {
              onChange(value);
            }
          }
        }}
      />
      <Button
        icon
        noPadding
        onClick={() => onIncrement(count < max ? count + 1 : count)}
      >
        <PlusSvg />
      </Button>
    </Wrapper>
  );
};

export default Counter;
