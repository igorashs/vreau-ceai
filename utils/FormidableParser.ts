import * as validator from '@/utils/validator';
import FormidableNameSpace, { Files, File, IncomingForm } from 'formidable';
import { NextApiRequest } from 'next';
import { Formidable, ProductFields } from 'types';
import { productMessages } from './validator/schemas/product';
import { NextApiSessionRequest } from './withSession';

type ParsedData = {
  fields: Partial<ProductFields>;
  files: Files & { src?: File };
};

type HandleProductForm = (
  req: NextApiRequest | NextApiSessionRequest,
  options: Partial<FormidableNameSpace.Options>,
  allowedFileTypes: string[],
) => Promise<ParsedData>;

const handleProductForm: HandleProductForm = async (
  req,
  options,
  allowedFileTypes,
) => {
  const data: ParsedData = await new Promise((resolve, reject) => {
    const form: Formidable = new IncomingForm(options);

    form.onPart = (part) => {
      if (part.filename) {
        // get file extension
        const fileType = part.mime?.split('/').pop() ?? '';

        // validate allowed file types
        if (!allowedFileTypes.includes(fileType)) {
          form.emit(
            'error',
            validator.createValidationError({
              message: productMessages.src.invalid,
              key: 'src',
            }),
          );
        }
      }

      form.handlePart(part);
    };

    form.parse(req, (err, fields, files) => {
      if (err) {
        // custom error on maxFileSize
        if (err.message.includes('maxFileSize')) {
          reject(
            validator.createValidationError({
              message: productMessages.src.max,
              key: 'src',
            }),
          );
        } else {
          reject(err);
        }
      }

      resolve({ fields, files });
    });
  });

  return data;
};

export default {
  handleProductForm,
};
