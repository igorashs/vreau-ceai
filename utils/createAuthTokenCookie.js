import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createAuthTokenCookie = (payload) => {
  const authToken = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  return cookie.serialize('auth-token', authToken, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });
};
