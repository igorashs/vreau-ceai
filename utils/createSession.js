import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createSession = (payload) => {
  const token = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  const [header, jwtPayload, signature] = token.split('.');
  const session = header + '.' + jwtPayload;

  const sessionCookie = cookie.serialize('session', session, {
    secure: process.env.NODE_ENV !== 'development',
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  const authCookie = cookie.serialize('auth', signature, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  return [sessionCookie, authCookie];
};
