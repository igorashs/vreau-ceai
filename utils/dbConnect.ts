import mongoose from 'mongoose';

const dbConnect = async () => {
  if (mongoose.connection.readyState >= 1) {
    return;
  }

  await mongoose.connect(process.env.MONGODB_URI as string, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    useFindAndModify: false,
  });
};

export default dbConnect;
