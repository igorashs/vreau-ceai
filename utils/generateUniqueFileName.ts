import { nanoid } from 'nanoid';

const generateUniqueFileName = (fileName: string) => {
  // grab file extension
  const matches = fileName?.match(/\..+$/);

  // create new file name
  const uniqueFileName = nanoid() + (matches && matches[0]);

  return uniqueFileName;
};

export default generateUniqueFileName;
