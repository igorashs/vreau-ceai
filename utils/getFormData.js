export const getFormData = (data) => {
  const formData = new FormData();

  for (const name in data) {
    if (name === 'src') continue;

    formData.append(name, data[name]);
  }

  if (data.src[0]) {
    formData.append('src', data.src[0]);
  }

  return formData;
};
