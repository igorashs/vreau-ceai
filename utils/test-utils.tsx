import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
