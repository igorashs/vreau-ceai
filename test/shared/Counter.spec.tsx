import { fireEvent, render } from '@/utils/test-utils';
import Counter from '@/shared/Counter';

describe('Counter', () => {
  it('renders correctly', () => {
    const onChangeMock = jest.fn();
    const onSelectMock = jest.fn();

    const { container, getByRole } = render(
      <Counter onChange={onChangeMock} />,
    );

    const counterInput = getByRole('spinbutton');
    counterInput.onselect = onSelectMock;

    fireEvent.focus(counterInput);

    expect(onSelectMock).toBeCalledTimes(1);
    expect(counterInput).toHaveValue(0);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('.onChange', () => {
    it('returns changed value', () => {
      const onChangeMock = jest.fn();

      const { getByRole } = render(<Counter onChange={onChangeMock} />);

      const counterInput = getByRole('spinbutton');
      fireEvent.change(counterInput, { target: { value: '23' } });

      expect(onChangeMock).toBeCalledTimes(1);
      expect(onChangeMock).toBeCalledWith(23);
      expect(counterInput).toHaveValue(0);
    });
  });

  describe('min/max props limits', () => {
    it('remains between limits', () => {
      const onChangeMock = jest.fn();
      const min = 1;
      const max = 10;

      const { getByRole } = render(
        <Counter onChange={onChangeMock} min={min} max={max} />,
      );

      const counterInput = getByRole('spinbutton');

      expect(counterInput).toHaveValue(min);
      fireEvent.change(counterInput, { target: { value: '0' } });
      expect(onChangeMock).toBeCalledWith(min);

      fireEvent.change(counterInput, { target: { value: '11' } });
      expect(onChangeMock).toBeCalledWith(max);

      expect(onChangeMock).toBeCalledTimes(2);
    });
  });

  describe('count prop', () => {
    it('sets the current Counter value', () => {
      const onChangeMock = jest.fn();
      const count = 23;

      const { getByRole } = render(
        <Counter onChange={onChangeMock} count={count} />,
      );

      expect(getByRole('spinbutton')).toHaveValue(count);
    });

    it('respects min/max limits', () => {
      const onChangeMock = jest.fn();
      const min = 1;
      const max = 10;

      const { getByRole, rerender } = render(
        <Counter onChange={onChangeMock} min={min} max={max} count={0} />,
      );

      const counterInput = getByRole('spinbutton');
      expect(counterInput).toHaveValue(min);

      rerender(
        <Counter onChange={onChangeMock} min={min} max={max} count={23} />,
      );

      expect(counterInput).toHaveValue(max);
    });
  });

  describe('.onDecrement', () => {
    it('has onChange default handler', () => {
      const onChangeMock = jest.fn();

      const { getAllByRole } = render(
        <Counter onChange={onChangeMock} count={4} />,
      );

      const [decrementBtn] = getAllByRole('button');
      fireEvent.click(decrementBtn);

      expect(onChangeMock).toBeCalledTimes(1);
      expect(onChangeMock).toBeCalledWith(3);
    });

    it('has its own handler', () => {
      const onChangeMock = jest.fn();
      const onDecrementMock = jest.fn();

      const { getAllByRole } = render(
        <Counter
          onChange={onChangeMock}
          count={4}
          onDecrement={onDecrementMock}
        />,
      );

      const [decrementBtn] = getAllByRole('button');
      fireEvent.click(decrementBtn);

      expect(onChangeMock).toBeCalledTimes(0);
      expect(onDecrementMock).toBeCalledTimes(1);
      expect(onDecrementMock).toBeCalledWith(3);
    });

    it('respects min/max limits', () => {
      const onChangeMock = jest.fn();
      const min = 2;
      const max = 10;

      const { getAllByRole, rerender } = render(
        <Counter onChange={onChangeMock} min={min} max={max} count={-2} />,
      );

      const [decrementBtn] = getAllByRole('button');
      fireEvent.click(decrementBtn);

      expect(onChangeMock).toBeCalledWith(min);

      rerender(
        <Counter onChange={onChangeMock} min={min} max={max} count={22} />,
      );
      fireEvent.click(decrementBtn);

      expect(onChangeMock).toBeCalledTimes(2);
      expect(onChangeMock).toBeCalledWith(max - 1);
    });
  });

  describe('.onIncrement', () => {
    it('has onChange default handler', () => {
      const onChangeMock = jest.fn();

      const { getAllByRole } = render(
        <Counter onChange={onChangeMock} count={4} />,
      );

      const [, incrementBtn] = getAllByRole('button');
      fireEvent.click(incrementBtn);

      expect(onChangeMock).toBeCalledTimes(1);
      expect(onChangeMock).toBeCalledWith(5);
    });

    it('has its own handler', () => {
      const onChangeMock = jest.fn();
      const onIncrementMock = jest.fn();

      const { getAllByRole } = render(
        <Counter
          onChange={onChangeMock}
          count={4}
          onIncrement={onIncrementMock}
        />,
      );

      const [, incrementBtn] = getAllByRole('button');
      fireEvent.click(incrementBtn);

      expect(onChangeMock).toBeCalledTimes(0);
      expect(onIncrementMock).toBeCalledTimes(1);
      expect(onIncrementMock).toBeCalledWith(5);
    });

    it('respects min/max limit', () => {
      const onChangeMock = jest.fn();
      const min = 2;
      const max = 10;

      const { getAllByRole, rerender } = render(
        <Counter onChange={onChangeMock} min={min} max={max} count={-2} />,
      );

      const [, incrementBtn] = getAllByRole('button');
      fireEvent.click(incrementBtn);

      expect(onChangeMock).toBeCalledWith(min + 1);

      rerender(
        <Counter onChange={onChangeMock} min={min} max={max} count={22} />,
      );

      fireEvent.click(incrementBtn);

      expect(onChangeMock).toBeCalledTimes(2);
      expect(onChangeMock).toBeCalledWith(max);
    });
  });
});
