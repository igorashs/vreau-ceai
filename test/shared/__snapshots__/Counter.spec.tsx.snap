// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter renders correctly 1`] = `
.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 7px;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  padding: 4px;
  box-shadow: none;
  padding: 0;
}

.c1:hover {
  background-color: var(--accent);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 7px;
}

.c2 {
  border: 0;
  width: calc(var(--baseline) * 2);
  background-color: transparent;
  color: var(--accent-text-dark);
  border-bottom: 1px solid var(--accent);
  text-align: center;
}

.c2::-webkit-outer-spin-button,
.c2::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c2[type='number'] {
  -moz-appearance: textfield;
}

<div
  class="c0"
>
  <button
    class="c1"
  >
    <svg />
  </button>
  <input
    class="c2"
    type="number"
    value="0"
  />
  <button
    class="c1"
  >
    <svg />
  </button>
</div>
`;
